nerfstudio/model_components/losses.py

```diff
@@ -308,3 +308,234 @@ def depth_loss(
+def compute_scale_and_shift(prediction: torch.Tensor, target: torch.Tensor, mask: torch.Tensor):
+    """
+    https://github.com/EPFL-VILAB/omnidata/blob/318a75569934737e67902f903531324d1f48ae8f/omnidata_tools/torch/losses/midas_loss.py#L8
+    Args:
+        prediction: predicted depth map
+        target: ground truth depth map
+        mask: mask of valid pixels
+    Returns:
+        scale and shift for depth prediction
+    """
+    # system matrix: A = [[a_00, a_01], [a_10, a_11]]
+    a_00 = torch.sum(mask * prediction * prediction, (1, 2))
+    a_01 = torch.sum(mask * prediction, (1, 2))
+    a_11 = torch.sum(mask, (1, 2))
+
+    # right hand side: b = [b_0, b_1]
+    b_0 = torch.sum(mask * prediction * target, (1, 2))
+    b_1 = torch.sum(mask * target, (1, 2))
+
+    # solution: x = A^-1 . b = [[a_11, -a_01], [-a_10, a_00]] / (a_00 * a_11 - a_01 * a_10) . b
+    x_0 = torch.zeros_like(b_0)
+    x_1 = torch.zeros_like(b_1)
+
+    det = a_00 * a_11 - a_01 * a_01
+    valid = det.nonzero()
+
+    x_0[valid] = (a_11[valid] * b_0[valid] - a_01[valid] * b_1[valid]) / det[valid]
+    x_1[valid] = (-a_01[valid] * b_0[valid] + a_00[valid] * b_1[valid]) / det[valid]
+
+    return x_0, x_1
+
+
+def reduction_batch_based(image_loss: torch.Tensor, M: torch.Tensor):
+    """
+    average of all valid pixels of the batch
+    Args:
+        image_loss: loss of each image
+        M: number of valid pixels of each image
+    Returns:
+        average of all valid pixels of the batch
+    """
+
+    # avoid division by 0 (if sum(M) = sum(sum(mask)) = 0: sum(image_loss) = 0)
+    divisor = torch.sum(M)
+    reduction = 0 if divisor == 0 else torch.sum(image_loss) / divisor
+
+    return reduction
+
+
+def reduction_image_based(image_loss, M):
+    """
+    mean of average of valid pixels of an image
+    Args:
+        image_loss: loss of each image
+        M: number of valid pixels of each image
+    Returns:
+        mean of average of valid pixels of an image
+    """
+    # avoid division by 0 (if M = sum(mask) = 0: image_loss = 0)
+    valid = M.nonzero()
+
+    image_loss[valid] = image_loss[valid] / M[valid]
+
+    return torch.mean(image_loss)
+
+
+def mse_loss(prediction: torch.Tensor, target: torch.Tensor, mask: torch.Tensor, reduction=reduction_batch_based):
+    """
+    mse loss from omnidata repo
+    Args:
+        prediction: predicted depth map
+        target: ground truth depth map
+        mask: mask of valid pixels
+        reduction: reduction function, either reduction_batch_based or reduction_image_based
+    Returns:
+        mse loss based on reduction function
+    """
+    M = torch.sum(mask, (1, 2))
+    res = prediction - target
+    image_loss = torch.sum(mask * res * res, (1, 2))
+
+    return reduction(image_loss, 2 * M)
+
+
+def gradient_loss(prediction: torch.Tensor, target: torch.Tensor, mask: torch.Tensor, reduction=reduction_batch_based):
+    """
+    gradient loss from omnidata repo
```

Comment:
Do you have a link? Can you also provide more specific information about what this loss does.